Release choice

CarolineMorton opened the pull request:
## Background
We currently have two methods to release outputs to the outside world. 
1. Via Github
2. Via Output Publisher

At present we use the `-n` or `--new-publish` CLI flag to decide whether to do the new one or the old one. 
It would be better if this used a direct API call to the workspace status to decide which method should be used. 

## Implementation
This uses `requests` library to make a GET call to return the boolean value for `uses_new_release_flow`. There is then a OR statement added to the line that checks if `-n` has been passed in. This is to ensure that the CLI option still exists. 

## Testing
There are no tests for this at present but this is something we should add.

bloodearnest commented on publisher/release.py:
```diff
@@ -168,7 +170,12 @@ def release(options, release_dir):
-        if options.new_publish:
+        res = requests.get(
+            f"https://jobs.opensafely.org/api/v2/workspaces/{cfg['workspace']}/status"
+        )
```
I would add a `res.raise_for_status()` here to handle job-server errors.

bloodearnest commented on publisher/release.py:
```diff
@@ -168,7 +170,12 @@ def release(options, release_dir):
+        )
+        uses_new_workflow = res.json()
+
+        if options.new_publish or uses_new_workflow["uses_new_release_flow"]:
```
I might handle these cases separately, to be able to provide useful output to the user.

e.g.

````
new_workflow = uses_new_workflow["users_new_release_flow"]

if new_workflow:
     print("Publishing to job-server")
elif options.new_publish:
     print("Publishing to job-server, even though project currently uses github workflow.")
     new_workflow = True
else:
     print(f"Publishing to {repo}")

if new_workflow:
    ...
else:
    ...

```

ghickman commented on publisher/release.py:
```diff
@@ -168,7 +170,12 @@ def release(options, release_dir):
+        res = requests.get(
+            f"https://jobs.opensafely.org/api/v2/workspaces/{cfg['workspace']}/status"
+        )
+        uses_new_workflow = res.json()
```
Probably best to pull the boolean you care about out of the JSON here, it's not that obvious that the variable is going to contain a dictionary over a boolean.

ghickman reviewed (commented):
Great start!

CarolineMorton commented:
Thanks for all your feedback. 

I have added tests for `check_status() `which checks the API as discussed above. I refactored it out into its own function. 

I have spent some time considering if we could easily test the `release.release` workflow to see if it deals with the job server API call appropriately. I ran into a couple of issues with fiddling around with this and would be great to have your feedback to see if 
1. This is necessary - and if not, abandon
2. If it is necessary, then some more guidance on this. 

This is what I have tried so far. 

**The Attempt**
Writing a test function that tests the whole release.release workflow. I borrowed the `default_config` and `write_manifest` from `test_config.py` to get a manifest and a default configuration. This is messy though  - (they could be moved into conftest and utils respectively though)

This is what I wrote: 
```
@responses.activate
def test_release_use_new_workflow(options, tmp_path, default_config, urlopen):
    responses.add(responses.GET,
                  'https://jobs.opensafely.org/api/v2/workspaces/workspace/status',
                  json={'uses_new_release_flow': 'True'}, status=200)

    write_manifest(tmp_path)
    f = tmp_path / "file.txt"
    f.write_text("test")
    options.files = [str(f)]
    options.new_publish = False
    options.yes = True

    urlopen.add_response(
        HTTPStatus.CREATED, headers={"Location": "http://hatch/release/id"}
    )
    urlopen.add_response(HTTPStatus.CREATED)
    urlopen.add_response(HTTPStatus.CREATED)
    urlopen.add_response(HTTPStatus.CREATED)

    release.release(
        options=options,
        release_dir=tmp_path
    )

    assert len(responses.calls) == 1
```
This is supposed to test the options of `--new-publish` CLI flag not being passed in but the workspace exists with the `uses_new_release_flow` in the Job Server. 

When `options.new_publish` is set to `True` this works fine, and gets around the sticky issue of `/release` not being a thing for made up workspace by reusing the urlopen work from `test_upload`. 

The above code does fail however because of the check for [`workspace_status`](https://github.com/opensafely-core/output-publisher/blob/b3a23f9a8e904d66564f59d2a8ee21d2032c8145/publisher/config.py#L171) in `config.py`. This seems expected because the response is not being mocked (or responses lib doing its magic) but the bigger question, I had was what is it doing there 😔 It seems that it is a previous partial implementation of what I have done in release i.e. [checking the status ](https://github.com/opensafely-core/output-publisher/blob/b3a23f9a8e904d66564f59d2a8ee21d2032c8145/publisher/config.py#L11). 
If I comment out these two lines, my code above works fine and tests pass. 

So my question is (assuming we probably don't want two separate API calls checking the same thing - one from config.py and one from release.py):
1. Could we do what we want to do entirely within the config.py as it seems to at least partially have been implemented here first and perhaps that is a better place for it?
2. Shall I just cut out the `workspace_status` check from `config.py` entirely and move everything, as above to `release.py`. 

Hope that makes sense. Am off tomorrow so wanted to get this down in one paper (well, screen) as I _will_ forget by next week.

ghickman commented on publisher/release.py:
```diff
@@ -162,7 +162,8 @@ def check_status(workspace_name):
     uses_new_workflow = res.json()
-    return uses_new_workflow["uses_new_release_flow"]
+    res.raise_for_status()
+    return bool(uses_new_workflow["uses_new_release_flow"])
```
This should be a boolean already 🙂

ghickman commented on tests/test_release.py:
```diff
@@ -92,3 +94,30 @@ def test_releaseno_args(tmp_path):
+    responses.add(
+        responses.GET,
+        "https://jobs.opensafely.org/api/v2/workspaces/test_workspace/status",
+        json={"uses_new_release_flow": "True"},
```
You're passing a dictionary to the `json` kwarg (rather than actual JSON!), so a better way to test this would be to pass the literal `True`, which matches how the job-server endpoint returns this key.

ghickman reviewed (changes requested):
Looking good 👍 just one small tweak to the type of data we pass around.

CarolineMorton commented:
After discussion with @bloodearnest, we decided to move the code back into `config.py` so that all _decisions_ about the release progress are made in the config and passed into the release.py, rather than being split between the two. 

I have tried to surface some sensible errors if the API call to `/status` does not work, in particular if job server is down for some reason. I have added some more tests which account for options including:
- new_publish is False as arg but `uses_new_release_flow` is True on API call. SHould use new release. 
- new_publish is True but `uses_new_release_flow` is False on API call. Should use old release method. 
- Job Server down is down - should return error with Job Server down.

bloodearnest commented on tests/test_release.py:
```diff
@@ -15,6 +17,13 @@
 # a `study_repo` is an empty git repo
 
 
+@pytest.fixture
```
I would probably do the following rather than redefining it:

`from tests.config import urlopen`

CarolineMorton replied:
It seems i no longer need this at all so I have deleted it.

bloodearnest commented on tests/test_config.py:
```diff
@@ -141,9 +158,14 @@ def test_load_config_new_publish_no_files(options, default_config, tmp_path):
-):
+@responses.activate
+def test_load_config_old_publish_no_private_token(options, tmp_path):
+    responses.add(
```
Since this is so common, I would recommend defining a fixture for it, e.g.

```
@pytest.fixture
def old_release_flow():
    responses.add(
        ...
    )
```

And then using it for each of the `*_old_publish_*` tests?

CarolineMorton replied:
Thanks. I have made this a `pytest.yield_fixture` to make the responses still mock the request response. Using `pytest.fixture` directly causes an error as it cannot response to an API call in the appropriate way.

bloodearnest reviewed (commented):
Looks good to me, I think the error handling works

Left a couple of optional suggestions if you think they are a good idea.

CarolineMorton commented:
> > * new_publish is False as arg but `uses_new_release_flow` is True on API call. SHould use new release.
> > * new_publish is True but `uses_new_release_flow` is False on API call. Should use old release method.
> 
> Huh, this is opposite to what I think the code does. If `options.new_publish` is True, then it will use the _new_ method, and not call the status api, no?

Whoops, yes you are quite correct, and that was what I intended....!

ghickman commented on publisher/config.py:
```diff
@@ -4,41 +4,23 @@
-            pass
-
-    raise Exception(f"Error: {response.status} response from {url}: {error_msg}")
+    uses_new_workflow = res.json()
+    if res.status_code == 500:
```
I think this might blow up… `res.json()` will attempt parse the JSON string into a Python dictionary, but if the response contains content that isn't valid JSON I'm pretty sure this line will error.  Since it's more common to see that situation with a non-happy status code and we're nore using this value until after the status checks lets move it down to just before the `return`, you could even return from it directly with `return res.json()["uses_new_release_flow"]`.

CarolineMorton replied:
of course! now I can see that, its obvious! thanks George!

ghickman reviewed (changes requested):
Very nearly there, just one more bit, sorry!

ghickman reviewed (approved):
🎉
